photon-ml/src/test/scala/com/linkedin/photon/ml/function/ObjectiveFunctionTest.scala

```diff
@@ -16,343 +16,501 @@ package com.linkedin.photon.ml.function
+      })
+      .toList
 
-  def checkGradient(prefix: String, objectiveBefore: Double, objectiveAfter: Double, expected: Double): Unit = {
-    val numericDeriv = (objectiveAfter - objectiveBefore) / (2.0 * DERIVATIVE_DELTA)
-    val relativeErrorNum = Math.abs(numericDeriv - expected)
-    val relativeErrorFactor = Math.min(Math.abs(numericDeriv), Math.abs(expected))
-    val relativeErrorDen = if (relativeErrorFactor > 0) {
-      relativeErrorFactor
-    } else {
-      1
-    }
-    val relativeError = relativeErrorNum / relativeErrorDen
-
-    assertTrue(java.lang.Double.isFinite(objectiveBefore), s"Objective before step [$objectiveBefore] should be finite")
-    assertTrue(java.lang.Double.isFinite(objectiveAfter), s"Objective after step [$objectiveAfter] should be finite")
-
-    if (!(numericDeriv.isInfinite || numericDeriv.isNaN)) {
-      assertTrue(relativeError < GRADIENT_TOLERANCE ||
-          relativeErrorNum < GRADIENT_TOLERANCE,
-        "Computed gradient and numerical differentiation estimate should be close." +
-            s"$prefix estimated [$numericDeriv] " +
-            s"v. computed [$expected] with absolute error [$relativeErrorNum] and " +
-            s"relative error [$relativeError]")
-    }
+  /**
+    * Generate weighted outlier-dense data sets for Poisson regression.
+    *
+    * @return A List of [[LabeledPoint]]
+    */
+  def generateWeightedOutlierDataSetPoissonRegression(): List[LabeledPoint] = {
+    val r = new Random(WEIGHT_RANDOM_SEED)
+    drawSampleFromOutlierDenseFeaturesForPoissonRegressionLocal(
+      DATA_RANDOM_SEED,
+      TRAINING_SAMPLES,
+      PROBLEM_DIMENSION)
+      .map({ obj =>
+        assertEquals(obj._2.length, PROBLEM_DIMENSION, "Samples should have expected lengths")
+        val weight = r.nextDouble() * WEIGHT_RANDOM_MAX
+        new LabeledPoint(label = obj._1, features = obj._2, weight = weight)
+      })
+      .toList
   }
 
-  def checkHessian(prefix: String, gradBefore: Double, gradAfter: Double, expected: Double): Unit = {
-    val numericDeriv = (gradAfter - gradBefore) / (2.0 * DERIVATIVE_DELTA)
-    val relativeErrorNum = Math.abs(numericDeriv - expected)
-    val relativeErrorFactor = Math.min(Math.abs(numericDeriv), Math.abs(expected))
-    val relativeErrorDen = if (relativeErrorFactor > 0) {
+  /**
+    * Check a computed derivative value against a numerical estimate and throw an error is either the relative or
+    * absolute error is too great.
+    *
+    * @param prefix An informative message about the derivative being checked
+    * @param descriptor A description of the context for the numerical estimation
+    * @param deltaBefore The function value at delta units before the computation
+    * @param deltaAfter The function value at delta units after the computation
+    * @param expected The computed value of the derivative
+    * @param tolerance The relative & absolute error tolerance
+    */
+  def checkDerivativeError(
```

Comment:
@joshvfleming 
Fused 'checkGradient' and 'checkHessian' functions into one. They were both estimating the derivative in the same way.